File: Shashank_Code/SLBFGS.m
function [evals, ngz, stepsizes] = SLBFGS(init, s, y, rho,...
                                 m, Ng, Nh, M, func,...
                                 grad, hess, grank, hrank)
%   init--initial guess
%   m--limited memory constant (default is 5)
%   Ng--batch size for gradient
%   Nh--batch size for hessian
%   S,Y--nonempty dim by m matrices whose columns are s_i and y_i
%   Rho--dim by 1 matrix whose entries are 1/s_i'y_i
%   grad--computing gradient
%   hess--computing hessian 
%   grank--rank of gradient
%   hrank--rank of hessian
%   M--frequency of hessian update
%   stepsizes--stepsizes 
gam = 0.9; % line search step factor
jmax = ceil(log(1e-14)/log(gam)); % max # of iterations in line search
eta = 0.1; % backtracking stopping criterion factor
xi_g = randi(grank, Ng, 1);
g = grad(xi_g,init);
tol = 1e-10;
nor = norm(g);
iter = 1;
x = init;
evals = func(x);
ngz = nor;
stepsizes = 1;
Ng = min(Ng,64);
Nh = min(Nh,64);
NN = 15;
while ((nor > tol) && (iter < 1e4))
    % We compute the direction via the LBFGS routine 
    if iter < m 
        I = 1 : iter;
        p = finddirection(g,s(:,I),y(:,I),rho(I));
    else
        p = finddirection(g,s,y,rho);
    end
    % We compute the step size via ls 
    [a,j] = ls(x,p,g,func,eta,gam,jmax);
    if j == jmax
        p = -g;
        [a,j] = ls(x,p,g,func,eta,gam,jmax);
    end
    % 
    step = a*p;
    xnew = x + step; %xnew is x_k+1; x is x_k 
     %snew is sk_k+1 
    %Stochastic step for Hessian 
    if mod(iter,M) == 0
        xi_h = randperm(hrank, Nh);
        ynew = hess(xi_h, xnew, snew);
                                             
        for k=2:100
                xi_h = randperm(hrank, Nh);
                yz = hess(xi_h, xnew, snew); 
                ynew = yz + ynew; 
        end
        ynew = 0.01*ynew;
        snew = step;
        % updating s, y and rho
        s = [s snew];
        y = [y ynew];
        rho = [rho rhonew];
    end
    
    
    if size(s,2) > m
        s = s(:, 2:end);
        y = y(:, 2:end);
        rho = rho(:,2:end);
    end
    % preparing for next iteration 
    x = xnew; %x_k+1
    xi_g = randi(grank, Ng, 1); 
    g = grad(xi_g,x); 
    for i=2:100
        xi_g = randi(grank, Ng, 1); 
        g = g + grad(xi_g,x);
    end% Random approximation to the gradient at the next
    g = 0.01*g;                  % iterate
    nor = norm(g);
    evals = [evals func(x)];
    ngz = [ngz nor];
    stepsizes = [stepsizes a];
    iter = iter + 1;
 end
sstep = 1;
N = 5;
NN = 10;
% for ii = 1:NN
%     ss = sstep/(2^ii);
%     nsteps = ceil(N*2^ii/ii);
%                 for i = 1 : nsteps
%                     if iter < m 
%                         I = 1 : iter;
%                         p = finddirection(g,s(:,I),y(:,I),rho(I));
%                     else
%                         p = finddirection(g,s,y,rho);
%                     end
%                     
%                     step = ss*p;
%                     xnew = x + step; %xnew is x_k+1; x is x_k 
%                     snew = step; %snew is sk_k+1 
%                     
%                     %Stochastic step for Hessian 
%                     if mod(iter,M) == 0
%                         xi_h = randperm(hrank, Nh);
%                         ynew = hess(xi_h, xnew, snew);
%                                              
%                         for k=2:100
%                             xi_h = randperm(hrank, Nh);
%                             yz = hess(xi_h, xnew, snew); 
%                             ynew = yz + ynew; 
%                         end
%                         ynew = 0.01*ynew;
%                         
%                     else
%                         xi_g = randperm(grank, Ng);
%                         gnew = grad(xi_g, xnew);
%                         for k=2:100
%                             xi_g = randperm(grank, Ng);
%                             gz = grad(xi_g, xnew);
%                             gnew = gnew + gz;
%                         end
%                         gnew = 0.01*gnew;
%                         ynew = gnew - g;
%                         snew = xnew - x;
%                         rhonew = 1/((snew(:,1)')*ynew(:,1));
%                     end
%                     % updating s, y and rho
%                     s = [s snew];
%                     y = [y ynew];
%                     rho = [rho rhonew];
%                     if size(s,2) > m
%                         s = s(:, 2:end);
%                         y = y(:, 2:end);
%                         rho = rho(:,2:end);
%                     end
%                     
%                     % preparing for next iteration 
%                     x = xnew; %x_k+1
%                     g = gnew; % Random approximation to the gradient at the next
%                                       % iterate
%                     nor = norm(g);
%                     evals = [evals func(x)];
%                     ngz = [ngz nor];
%                     stepsizes = [stepsizes ss];
%                     iter = iter + 1;
%                 end
% end

end

File: Alex_Code/SLBFGS.py
import numpy as np
from finddirection import finddirection
from linesearch import linesearch
import time

def SLBFGS(fun, gfun, Y, x0, stepmethod = 'linesearch', m = 5, stepsperHupdate = 10, batchsizeg = 64, batchsizeH = 100, maxiter = 500, Hevalmethod = 'onestep'):
    # Stochastic L-BFGS routing for computing minimum
    #  fun - loss function
    #  gfun - loss function gradient
    #  Y - Data array
    #  x0 - initial guess for weights
    #  stepmethod - method for computing step size, default is linesearch, otherwise is set to a function 
    #  m - size of memory for Hessian approximation
    #  stepsperHupdate - number of iterations between each Hessian update
    #  batchsizeg/H - batchsizes for computation of gradient/Hessian
    #  maxiter - maximum number of iterations
    #  Hevalmethod - method for computing H update, can either be over onestep or stepsperHupdate
    
    n = Y.shape[0]
    # Set parameters for line search
    gam = 0.9
    eta = 0.5
    tol = 1e-10
    jmax = round(np.ceil(np.log(1e-14)/(np.log(gam))))
    
    # Initialize arrays for storing s, y, rho, f, gradnorm, x at each iteration, and runtime
    s = np.zeros((x0.size, m))
    y = np.zeros((x0.size, m))
    rho = np.zeros(m)
    f = np.zeros(maxiter+1)
    gradnorm  = np.zeros(maxiter+1)
    xiter = np.zeros((maxiter+1,x0.size))
    runtime = np.zeros(maxiter+1)
    tic = time.perf_counter()
    Iall = np.arange(n) #Array for all indices
    
    # Evaluate initial function values
    f[0] = fun(Iall,Y,x0)
    xiter[0] = x0
    bszg = np.minimum(batchsizeg, n) #Set batchsizes to max if batchsizes are too large for data
    bszH = np.minimum(batchsizeH, n)
    
    # Set initial samples for gradient and Hessian computation
    Ig = np.random.permutation(n)[:bszg]
    IH = np.random.permutation(n)[:bszH]
    g = gfun(Ig,Y,x0) # Compute grad for initial step
    gH = gfun(IH,Y,x0) # Compute grad for initial Hessian
    gradnorm[0] = np.linalg.norm(g)
    # Get initial step size
    if stepmethod is 'linesearch':
        linefun = lambda x: fun(Ig, Y, x)
        a, j = linesearch(x0, -g, g, linefun, eta, gam, jmax)
    else:
        a = stepmethod(0)
    # Compute first step
    xnew = x0 - a*g
    toc = time.perf_counter() # Store runtime
    runtime[0] = toc - tic
    gnew = gfun(Ig, Y, xnew) # Compute new gradient
    gnewH = gfun(IH, Y, xnew) # Compute new gradient for Hessian and store s, y and rho
    s[:,0] = xnew - x0
    y[:,0] = gnewH - gH
    rho[0] = 1/np.dot(s[:,0],y[:,0])
    # Begin next iteration and store values
    x = xnew
    g = gnew
    nor = np.linalg.norm(g)
    xiter[1] = x
    gradnorm[1] = nor
    f[1] = fun(Iall, Y,x)
    itr = 1
    mitr = 1
    # while nor > tol and itr < maxiter: # End iteration if norm(g) < tol
    while itr < maxiter:
        if mitr < m: #If we have yet to fill up s, y, and rho to size m
            p = finddirection(g, s[:,:mitr], y[:,:mitr],rho[:mitr])
        else:
            p = finddirection(g,s,y,rho)
        # Find step size
        if stepmethod is 'linesearch':
            linefun = lambda x: fun(Ig, Y, x)
            a,j = linesearch(x,p,g,linefun,eta,gam,jmax)
            if j == jmax:
                p = -g
                a,j = linesearch(x,p,g,linefun,eta,gam,jmax)
        else:
            a = stepmethod(itr)
        step = a*p # Compute next step
        xnew = x + step
        xiter[itr + 1] = xnew
        toc = time.perf_counter()
        runtime[itr] = toc - tic
        f[itr + 1] = fun(Iall,Y,xnew)
        if itr % stepsperHupdate == 0: # Every stepsperHupdate iterations, we update H
            IH = np.random.permutation(n)[:bszH] # Get a new, larger sample
            # Evaluate H either over one step or multiple
            if Hevalmethod is 'onestep':
                snew = step
                ynew = gfun(IH, Y, xnew) - gfun(IH, Y, x)
            else:
                snew = xnew - xiter[itr + 1 - stepsperHupdate]
                ynew = gfun(IH, Y, xnew) - gfun(IH,Y,xiter[itr + 1 - stepsperHupdate])
            # Add new value to s,y, and rho and remove oldest column
            s = np.append(snew.reshape((-1,1)), s[:,:m-1], axis = 1)
            y = np.append(ynew.reshape((-1,1)), y[:,:m-1], axis = 1)
            rho = np.append(1/np.dot(snew,ynew), rho[:m-1])
            mitr+=1
        x = xnew
        # Find new random sample and recalculate gradient
        Ig = np.random.permutation(n)[:bszg]
        g = gfun(Ig, Y, x)
        nor = np.linalg.norm(g)
        gradnorm[itr + 1] = nor
        itr += 1
    # If tolerance was reached, truncate results arrays
    if itr < maxiter:
        xiter = xiter[:itr+1]
        gradnorm = gradnorm[:itr+1]
        f = f[:itr+1]
        runtime = runtime[:itr+1]
       
    return xiter, f, gradnorm, runtime
        
        